x-pack/plugins/fleet/server/services/actions/utils.ts

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { KueryNode } from '@kbn/es-query';
+import { get } from 'lodash';
+
+const getFieldType = (
+  key: string | undefined | null,
+  indexMappings: FleetActionsIndexMapping
+): string => {
+  const mappingKey = `properties.${key}.type`;
+  return key != null ? get(indexMappings, mappingKey) : '';
+};
+
+export const isFieldDefined = (
+  indexMappings: FleetActionsIndexMapping | FleetActionsResultsIndexMapping,
+  key: string
+): boolean => {
+  const mappingKey = 'properties.' + key;
+  return typeof get(indexMappings, mappingKey) !== 'undefined';
+};
+
+export const hasFieldKeyError = (
+  key: string | null | undefined,
+  fieldTypes: string[],
+  indexMapping: FleetActionsIndexMapping | FleetActionsResultsIndexMapping,
+  indexType: IndexType
+): string | null => {
+  const allowedKeys = Object.keys(indexMapping.properties).join();
+  if (key == null) {
+    return `The key is empty and should be one of [${allowedKeys}]`;
+  }
+  if (key) {
+    const allowedFieldTypes = fieldTypes.every((type) =>
+      indexType === 'actions'
+        ? ALLOWED_FLEET_ACTIONS_FIELD_TYPES.includes(type)
+        : ALLOWED_FLEET_ACTIONS_RESULTS_FIELD_TYPES.includes(type)
+    );
+    const indexName = indexType === 'actions' ? '.fleet-actions' : '.fleet-actions-results';
+    if (!isFieldDefined(indexMapping, key)) {
+      return `This key '${key}' does not exist in ${indexName} index mappings`;
+    }
+    if (!allowedFieldTypes) {
+      return `This key '${key}' does not match allowed field types in ${indexName} index mappings`;
+    }
+  }
+  return null;
+};
+
+export const ALLOWED_FLEET_ACTIONS_RESULTS_FIELD_TYPES = ['keyword'];
+export const ALLOWED_FLEET_ACTIONS_FIELD_TYPES = ['keyword', 'date'];
+
+export interface FleetActionsIndexMapping {
+  properties: {
+    action_id: {
+      type: 'keyword';
+    };
+    agents: {
+      type: 'keyword';
+    };
+    input_type: {
+      type: 'keyword';
+    };
+    '@timestamp': {
+      type: 'date';
+    };
+    type: {
+      type: 'keyword';
+    };
+    user_id: {
+      type: 'keyword';
+    };
+  };
+}
+
+export interface FleetActionsResultsIndexMapping {
+  properties: {
+    action_id: {
+      type: 'keyword';
+    };
+    agent_id: {
+      type: 'keyword';
+    };
+  };
+}
+
+export const allowedFleetActionsFields = Object.freeze({
```

Comment:
Suggestion: for deep object structures, use `deepFreeze()` from `import { deepFreeze } from '@kbn/std';`. `Object.freeze()` only does a shallow freeze

_(same comment for others below)_

Reply:
done 7b2bb27ae9f94cc74442498a55b28a8f407c692c